### PR TITLE
feat: add support multi-ranges

### PR DESCRIPTION
**Pull request by ustuzhanin**
There is three breaking changes:
1. ETCD store lock name
2. Name of kubernetes resource for pool
3. Replace ip offset in CRD to IP strings

### CODE REVIEW

**Comment by ustuzhanin** [comment]
@dougbtv

**Comment by dougbtv** [comment]
Thanks a ton for the contribution! This will take a bit of time to dig through and think through for sure. So give me a little time, but, thanks a bunch for putting this together, know that I appreciate it.

File: pkg/storage/storage.go [hunk 1]
```diff
@@ -66,9 +66,43 @@ func IPManagement(mode int, ipamConf types.IPAMConfig, containerID string) (net.
...
+		logging.Debugf("try ipRange:%+v", ipRange)
+		newip, err = IPManagementRange(ctx, mode, ipRange, ipam, containerID)
+		logging.Debugf("result allocate in range %+v IP:%+v err:%v", ipRange, newip, err)
+		if err == nil {
+			gateway = ipRange.Gateway
+			break
+
+		} else {
+			switch mode {
+			case 0:
```
Comment (crandles) [inline pkg/storage/storage.go:83]: can you use the constants `types.Allocate` and `types.Deallocate` instead of 0/1?

https://github.com/dougbtv/whereabouts/blob/master/pkg/types/types.go

File: pkg/storage/kubernetes.go [hunk 2]
```diff
@@ -58,7 +52,7 @@ func NewKubernetesIPAM(containerID string, ipamConf whereaboutstypes.IPAMConfig)
 	if err != nil {
 		return nil, err
 	}
-	return &KubernetesIPAM{c, ipamConf, containerID, namespace, DatastoreRetries}, nil
+	return &KubernetesIPAM{c, ipamConf, containerID, namespace, ipamConf.Name, DatastoreRetries}, nil
```
Comment (crandles) [inline pkg/storage/kubernetes.go:55]: Interesting, so instead of using the ip range as the resource name, this is utilizing the `"name"` of the CNI config (e.g. `mynet` from the tests). 

I'm not against this, we're rethinking the pool resource type in #51; this is a step in that direction I believe.

File: pkg/storage/storage.go (hunk 1 above)
Comment (ustuzhanin) [reply pkg/storage/storage.go]: fixed

**Comment by ustuzhanin** [comment]
And what will be the solution? We have already implemented it - it works

**Comment by ustuzhanin** [comment]
Hi :) We have been using this patch in production since October 20th. It works. Will there be movements?

**Comment by dionysius** [comment]
Allow me to hop in. We'd really like to see this feature. This all started when I asked for this feature in https://github.com/openshift/whereabouts-cni/issues/13.

Allow me to copy the issue description here:

-----

There seems to be no support for multiple ranges (or at least not documented). It seems this *very crucial* part of the host-local plugin has not been picked up.

### Use case
A /24 range is used up, since the cluster grew and grew. You could want to go /23, but the next /24 range is already reserved or you can't use the same gateway as the *actual* network is a /24 and the other half wouldn't be able to reach the gateway. So it's **technically not possible** to extend the range!

### Proposal
Introduce a `ranges` property which accepts a slice of `range-item`. A `range-item` consists of: `range`, `range_start`, `range_end`, `exclude`, `gateway` and as well the properties from the static plugin `dns`, `routes` and `addresses` as you already do.

So basically the the same properties you have as right now, but allowing to put them into a slice.

### Compatibility
Parsing of the current format would still be valid. Operational code would be changed to work with the `ranges` slice. After parsing the config the values of these fields created a new `range-item` and are added to the `ranges` slice. And since we changed the operational code to to work with the `ranges` slice it is already ready to be applied.

### Otherwise
Once the range is used up, you have to migrate **everything** to a completely new cluster with a new but bigger range, but also still limited until you find yourself in this situation again.

**Review by e0ne — CHANGES_REQUESTED** [review]

File: pkg/allocate/allocate.go [hunk 3]
```diff
@@ -35,7 +43,7 @@ func AssignIP(ipamConf types.IPAMConfig, reservelist []types.IPReservation, cont
 }
 
 // DeallocateIP assigns an IP using a range and a reserve list.
-func DeallocateIP(iprange string, reservelist []types.IPReservation, containerID string) ([]types.IPReservation, error) {
+func DeallocateIP(_ types.IPRange, reservelist []types.IPReservation, containerID string) ([]types.IPReservation, error) {
```
[the remark this answers is not in the document]
Comment (ustuzhanin) [reply pkg/allocate/allocate.go]: fixed

**Comment by ustuzhanin** [comment]
i am merge upstream, conflicts reloved.
need any more?
